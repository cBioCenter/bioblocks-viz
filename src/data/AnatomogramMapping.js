"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnatomogramMapping = {
    homo_sapiens: {
        P11A: ['UBERON_0000178'],
        P11B: ['UBERON_0001155'],
        P12A: ['UBERON_0000955'],
        P9A: ['UBERON_0000977'],
        UBERON_0000178: ['P11A'],
        UBERON_0000955: ['P12A'],
        UBERON_0000977: ['P9A'],
        UBERON_0001155: ['P11B'],
    },
    mus_musculus: {
        UBERON_0000948: ['Heart', 'Heart_and_Aorta', 'Heart_and_Aorta-10X_P7_4'],
        UBERON_0001255: ['Bladder', 'Bladder-10X_P4_3', 'Bladder-10X_P4_4', 'Bladder-10X_P7_7'],
        UBERON_0001723: ['Tongue', 'Tongue-10X_P4_0', 'Tongue-10X_P4_1', 'Tongue-10X_P7_10'],
        UBERON_0001911: ['Mammary', 'Mammary_Gland', 'Mammary_Gland-10X_P7_12', 'Mammary_Gland-10X_P7_13'],
        UBERON_0002048: ['Lung', 'Lung-10X_P7_8', 'Lung-10X_P7_9', 'Lung-10X_P8_12', 'Lung-10X_P8_13'],
        UBERON_0002106: ['Spleen', 'Spleen-10X_P4_7', 'Spleen-10X_P7_6'],
        UBERON_0002107: ['Liver', 'Liver-10X_P4_2', 'Liver-10X_P7_0', 'Liver-10X_P7_1'],
        UBERON_0002113: ['Kidney', 'Kidney-10X_P4_5', 'Kidney-10X_P4_6', 'Kidney-10X_P7_5'],
        UBERON_0002370: ['Thymus', 'Thymus-10X_P7_11'],
        UBERON_0002371: ['Marrow', 'Marrow-10X_P7_2', 'Marrow-10X_P7_3'],
        UBERON_0003126: ['Trachea', 'Trachea-10X_P8_14', 'Trachea-10X_P8_15'],
        UBERON_0014892: ['Muscle', 'Limb', 'Limb_Muscle', 'Limb_Muscle-10X_P7_14', 'Limb_Muscle-10X_P7_15'],
    },
};
//# sourceMappingURL=AnatomogramMapping.js.map