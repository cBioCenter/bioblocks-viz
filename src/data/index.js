"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./AminoAcid"), exports);
tslib_1.__exportStar(require("./Nucleotide"), exports);
tslib_1.__exportStar(require("./AnatomogramMapping"), exports);
tslib_1.__exportStar(require("./Bioblocks1DSection"), exports);
tslib_1.__exportStar(require("./BioblocksConfig"), exports);
tslib_1.__exportStar(require("./BioblocksData"), exports);
tslib_1.__exportStar(require("./BioblocksPDB"), exports);
tslib_1.__exportStar(require("./CodonTable"), exports);
tslib_1.__exportStar(require("./CouplingContainer"), exports);
tslib_1.__exportStar(require("./Dataset"), exports);
tslib_1.__exportStar(require("./IUPACData"), exports);
tslib_1.__exportStar(require("./Markers"), exports);
tslib_1.__exportStar(require("./PfamProtein"), exports);
tslib_1.__exportStar(require("./PlotlyData"), exports);
tslib_1.__exportStar(require("./Protein"), exports);
tslib_1.__exportStar(require("./Seq"), exports);
tslib_1.__exportStar(require("./SeqIO"), exports);
tslib_1.__exportStar(require("./SeqRecord"), exports);
tslib_1.__exportStar(require("./TintedBioblocks1DSection"), exports);
tslib_1.__exportStar(require("./event"), exports);
//# sourceMappingURL=index.js.map