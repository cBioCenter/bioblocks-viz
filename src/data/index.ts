export * from './AminoAcid';
export * from './Chell1DSection';
export * from './ChellConfig';
export * from './ChellData';
export * from './ChellPDB';
export * from './CouplingContainer';
export * from './PlotlyData';
export * from './Protein';
export * from './Spring';
export * from './TintedChell1DSection';

export * from './event';
