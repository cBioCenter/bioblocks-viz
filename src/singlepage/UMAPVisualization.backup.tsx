// import { fromJS, List } from 'immutable';
import * as Papa from 'papaparse';
import * as React from 'react';
import { UMAP } from 'umap-js';

// import { setTimeout } from 'timers';
// import { Seq } from 'immutable';
import { ComponentCard, defaultPlotlyLayout, PlotlyChart } from '~bioblocks-viz~/component';
// import { parsePath } from 'history';
// import { fetchMatrixData } from '~bioblocks-viz~/helper';

export interface IUMAPVisualizationProps {
  numIterationsBeforeReRender: number;
  numSequencesToShow: number;
}

export interface IUMAPVisualizationState {
  allSequencesFastaFile: null | FastaFile;
  selectedSequencesFastaFile: null | FastaFile;
  taxonomy: null | { [key: string]: { [key: string]: string } };
  twoDimensionalData: number[][];
}

export const fetchFastaFile = async (filename: string) => {
  const response = await fetch(filename);
  if (response.ok) {
    return FastaFile.fromFileText(await response.text());
  } else {
    throw new Error('error' + response);
  }
};

export interface SequenceEntry {
  title: string;
  aaSequence: string;
  binaryAASequence: number[];
}

export function hammingDistance(x: number[], y: number[]) {
  let result = 0;
  for (let i = 0; i < x.length; i++) {
    if (x[i] !== y[i]) {
      result += 1;
    }
  }

  return result;
}

export class FastaFile {
  public static getBinarySequenceFromAASequence(aaSequence: string) {
    return aaSequence.split('').reduce((binAccumulator, aaCode) => {
      const arr = new Array(21).fill(0);
      arr[FastaFile.aaList.indexOf(aaCode)] = 1;
      binAccumulator.push(...arr);

      return binAccumulator;
    }, new Array<number>());
  }
  public static fromFileText(fileText: string) {
    // parse fasta file line by line
    const sequences = fileText.split(/\r|\n/).reduce((accumulator, line) => {
      const trimmedLine = line.trim();
      if (trimmedLine.length > 0) {
        if (trimmedLine[0] === '>') {
          accumulator.push({
            aaSequence: '',
            binaryAASequence: [],
            title: trimmedLine.substr(1),
          });
        } else if (accumulator.length > 0) {
          const aaSequnceOnLine = trimmedLine.replace(/\./g, '-').toUpperCase();
          accumulator[accumulator.length - 1].aaSequence += aaSequnceOnLine;
          accumulator[accumulator.length - 1].binaryAASequence.push(
            ...FastaFile.getBinarySequenceFromAASequence(aaSequnceOnLine),
          );
        }
      }

      return accumulator;
    }, new Array<SequenceEntry>());

    return new FastaFile(sequences);
  }

  private static aaList = 'ACDEFGHIKLMNPQRSTVWY-';
  private sequences: SequenceEntry[];

  constructor(sequences: SequenceEntry[]) {
    this.sequences = sequences;
  }

  public getSequenceObjects() {
    return this.sequences;
  }

  public getBinarySequencesFlat() {
    return this.sequences.map(seqEntry => {
      return seqEntry.binaryAASequence;
    });
  }

  public getIntegerAASequences() {
    // Returns an array of integers for each sequence. The numbers
    // have no meaning other than to check for equality in a hamming
    // distance function.
    return this.sequences.map(seqEntry => {
      return seqEntry.aaSequence.split('').map(aaCode => {
        return FastaFile.aaList.indexOf(aaCode);
      });
    });
  }

  public getRandomSetOfSequences(n: number) {
    // randomly select "n" sequences from this FastaFile's sequences and
    // return a new FastaFile object with the new random set. If "n" is
    // larger than the number of sequences, return all sequences.
    if (n > this.sequences.length) {
      return new FastaFile([...this.sequences]);
    }
    let sequenceBowl = [...this.sequences];
    const randomSequences = new Array<SequenceEntry>();
    while (randomSequences.length < n) {
      // tslint:disable-next-line: insecure-random
      const randomIdx = Math.floor(Math.random() * sequenceBowl.length); // btw 0 and length of in sequenceBowl
      randomSequences.push(sequenceBowl[randomIdx]);
      sequenceBowl = sequenceBowl.reduce((accumulator, currentValue, currentIdx) => {
        if (currentIdx !== randomIdx) {
          accumulator.push(currentValue);
        }

        return accumulator;
      }, new Array<SequenceEntry>());
    }

    return new FastaFile(randomSequences);
  }
}

// tslint:disable-next-line: max-classes-per-file
export class UMAPVisualization extends React.Component<IUMAPVisualizationProps, IUMAPVisualizationState> {
  public static defaultProps = {
    numIterationsBeforeReRender: 1,
    numSequencesToShow: 5000,
  };

  private xRange: number[] = [-20, 20];
  private yRange: number[] = [-20, 20];
  private classColors = {
    asdf: 'asdf',
    32: 'asdf',
  };

  private phylumColors = {
    Actinobacteria: '#1f77b4',
    Alphaproteobacteria: '#ff7f0e',
    Bacilli: '#2ca02c',
    Betaproteobacteria: '#d62728',
    Clostridia: '#9467bd',
    Gammaproteobacteria: '#17becf',
  };

  // public componentDidUpdate(prevProps: IUMAPVisualizationProps, prevState: IUMAPVisualizationState) { }

  private counter = 0;

  constructor(props: IUMAPVisualizationProps) {
    super(props);
    this.state = {
      allSequencesFastaFile: null,
      selectedSequencesFastaFile: null,
      taxonomy: null,
      twoDimensionalData: new Array<number[]>(),
    };
  }

  public async componentDidMount() {
    const { numIterationsBeforeReRender, numSequencesToShow } = this.props;
    /*const fastaSequences = await fetchFastaFile(
      'datasets/betalactamase_alignment/7fa1c5691376beab198788a726917d48_b0.4.a2m',
    );*/

    let taxonomy: null | { [key: string]: { [key: string]: string } } = null;
    Papa.parse('datasets/betalactamase_alignment/PSE1_NATURAL_TAXONOMY.csv', {
      complete(results) {
        console.log('csv results', results);

        taxonomy = results.data.reduce(
          (accumulator, seqTaxonomy) => {
            accumulator[seqTaxonomy.seq_name] = {
              class: seqTaxonomy.class,
              phylum: seqTaxonomy.phylum,
            };

            return accumulator;
          },
          {} as { [key: string]: { [key: string]: string } },
        );
        // console.log('taxonomy');
        // console.log(taxonomy);
      },
      download: true,
      header: true,
    });

    // const fastaSequences = await fetchFastaFile(
    //  'datasets/AQP4_alignment/AQP4_HUMAN_eval2_region_22to255_raw_focus.fasta',
    // );
    const fastaSequences = await fetchFastaFile('datasets/betalactamase_alignment/PSE1_natural_top5K_subsample.a2m');
    console.log('all fasta sequences:', fastaSequences);

    const randomFastaSequences = fastaSequences.getRandomSetOfSequences(numSequencesToShow);
    console.log('random fasta sequences:', randomFastaSequences);

    const integerAASequences = randomFastaSequences.getIntegerAASequences(); // faster than binary
    const t0 = performance.now();
    /*
    const umap = new UMAP({ nComponents: 2, distanceFn: hammingDistance });
    umap.fitAsync(integerAASequences, epochNum => {
      console.log(epochNum + '::' + (performance.now() - t0) / 1000 + ' sec');
      this.setState({
        twoDimensionalData: umap.getEmbedding(),
      });
    });*/

    // WORKING FOR UMAP sequences dataset
    setTimeout(() => {
      const umap = new UMAP({ nComponents: 2, distanceFn: hammingDistance, minDist: 0.1, spread: 1 });

      const nEpochs = umap.initializeFit(integerAASequences);
      console.log(`UMAP wants to do ${nEpochs} epochs`);

      const stepUmapFn = (counter: number) => {
        if (counter % numIterationsBeforeReRender === 0 && counter <= nEpochs) {
          if (counter % 50 === 0) {
            console.log(counter + '::' + (performance.now() - t0) / 1000 + ' sec');
          }
          const embedding = umap.getEmbedding();
          if (counter === 0) {
            console.log('embedding:', embedding);
          }

          this.setState({
            allSequencesFastaFile: fastaSequences,
            selectedSequencesFastaFile: randomFastaSequences,
            taxonomy,
            twoDimensionalData: embedding,
          });
        }
        umap.step();
        if (counter <= nEpochs) {
          setTimeout(() => {
            stepUmapFn(counter + 1);
          });
        }
      };

      stepUmapFn(0);
    });

    /* // WORKING FOR UMAP scRNAseq dataset
    const { numIterationsBeforeReRender } = this.props;

    const data = await fetchMatrixData('datasets/hpc/full/tsne_matrix.csv');

    const { numIterationsBeforeReRender } = this.props;
    console.log(`UMAP got data (${data.length}, ${data[0].length})`);

    const umap = new UMAP({ nComponents: 2 });
    const nEpochs = umap.initializeFit(data);
    console.log(`UMAP wants to do ${nEpochs} epochs`);

    const stepUmapFn = (counter: number) => {
      if (counter % numIterationsBeforeReRender === 0 && counter <= nEpochs) {
        console.log(`UMAP on iteration ${counter}`);
        const embedding = umap.getEmbedding();
        this.setState({
          twoDimensionalData: embedding,
        });
      }
      umap.step();
      if (counter <= nEpochs) {
        setTimeout(() => {
          stepUmapFn(counter + 1);
        });
      }
    };

    setTimeout(() => {
      stepUmapFn(0);
    });*/
  }
  public render() {
    // console.log(`UMAP render with length ${data.length}:`, JSON.stringify(data));

    let plotType: string = 'scattergl';
    const dataX = new Float32Array(this.state.twoDimensionalData.map(datum => datum[0]));
    const dataY = new Float32Array(this.state.twoDimensionalData.map(datum => datum[1]));
    let dataZ: Float32Array | undefined;
    if (
      this.state.twoDimensionalData &&
      this.state.twoDimensionalData.length > 0 &&
      this.state.twoDimensionalData[0].length > 2
    ) {
      dataZ = new Float32Array(this.state.twoDimensionalData.map(datum => datum[3]));
      plotType = 'scatter3d';
    }

    let seqTitles = new Array<string>();
    if (this.state.selectedSequencesFastaFile) {
      seqTitles = this.state.selectedSequencesFastaFile.getSequenceObjects().map(seqObj => {
        return seqObj.title;
      });
    }

    let colors = new Array<string>();
    if (this.state.selectedSequencesFastaFile && this.state.taxonomy) {
      colors = this.state.selectedSequencesFastaFile.getSequenceObjects().map(seqObj => {
        if (this.state.taxonomy) {
          const taxProps = this.state.taxonomy[seqObj.title];
          const color = taxProps && taxProps.phylum ? this.getClassColor(taxProps.class) : null;
          if (color) {
            return color;
          }
        }

        return 'gray';
      });
    }

    // smart range setting
    if (this.xRange[0] > Math.min(...dataX)) {
      this.xRange[0] = Math.min(...dataX) - 5;
    }
    if (this.xRange[1] < Math.max(...dataX)) {
      this.xRange[1] = Math.max(...dataX) + 5;
    }
    if (this.yRange[0] > Math.min(...dataY)) {
      this.yRange[0] = Math.min(...dataY) - 5;
    }
    if (this.yRange[1] < Math.max(...dataY)) {
      this.yRange[1] = Math.max(...dataY) + 5;
    }

    return (
      <div>
        asdf
        <ComponentCard componentName={'UMAPVisualization'}>
          <PlotlyChart
            layout={{
              ...defaultPlotlyLayout,
              dragmode: 'zoom', // 'select',
              margin: {
                b: 20,
              },
              xaxis: { autorange: false, range: this.xRange },
              yaxis: { autorange: false, range: this.yRange },
            }}
            data={[
              {
                marker: {
                  color: colors,
                },
                mode: 'markers',
                text: seqTitles,
                type: 'scattergl', // scatter3d
                x: dataX,
                y: dataY,
                z: dataZ,
              },
            ]}
            showLoader={false}
          />
        </ComponentCard>
      </div>
    );
  }

  private getPhylumColor(phylumToFind: string | undefined): string | undefined {
    const phylumColors: { [key: string]: string } = {
      Actinobacteria: '#ff7f0e',
      Bacteroidetes: '#2ca02c',
      Firmicutes: '#8c564b',
      Proteobacteria: '#e377c2',
    };
    if (phylumToFind) {
      return phylumColors[phylumToFind];
    }

    return undefined;
  }

  private getClassColor(classToFind: string | undefined): string | undefined {
    const classColors: { [key: string]: string } = {
      Actinobacteria: '#ff7f0e',
      Alphaproteobacteria: '#1f77b4',
      Bacilli: '#2ca02c',
      Betaproteobacteria: '#d62728',
      Clostridia: '#9467bd',
      Gammaproteobacteria: '#17becf',
    };
    if (classToFind) {
      return classColors[classToFind];
    }
  }
}
