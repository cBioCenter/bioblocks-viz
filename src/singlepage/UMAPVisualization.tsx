import * as Papa from 'papaparse';
import * as React from 'react';
import { UMAP } from 'umap-js';

import { ComponentCard, defaultPlotlyLayout, PlotlyChart } from '~bioblocks-viz~/component';

export interface IUMAPVisualizationProps {
  numIterationsBeforeReRender: number;
  numSequencesToShow: number;
}

export interface IUMAPVisualizationState {
  allSequencesFastaFile: null | FastaFile;
  selectedSequencesFastaFile: null | FastaFile;
  taxonomy: null | { [key: string]: { [key: string]: string } };
  twoDimensionalData: number[][];
}

export const fetchFastaFile = async (filename: string) => {
  const response = await fetch(filename);
  if (response.ok) {
    return FastaFile.fromFileText(await response.text());
  } else {
    throw new Error('error' + response);
  }
};

export function hammingDistance(x: number[], y: number[]) {
  let result = 0;
  for (let i = 0; i < x.length; i++) {
    if (x[i] !== y[i]) {
      result += 1;
    }
  }

  return result;
}

// tslint:disable-next-line: max-classes-per-file
export class UMAPVisualization extends React.Component<IUMAPVisualizationProps, IUMAPVisualizationState> {
  public static defaultProps = {
    numIterationsBeforeReRender: 1,
    numSequencesToShow: 5000,
  };

  private xRange: number[] = [-20, 20];
  private yRange: number[] = [-20, 20];
  private classColors = {
    asdf: 'asdf',
    32: 'asdf',
  };

  private phylumColors = {
    Actinobacteria: '#1f77b4',
    Alphaproteobacteria: '#ff7f0e',
    Bacilli: '#2ca02c',
    Betaproteobacteria: '#d62728',
    Clostridia: '#9467bd',
    Gammaproteobacteria: '#17becf',
  };

  private counter = 0;

  constructor(props: IUMAPVisualizationProps) {
    super(props);
    this.state = {
      allSequencesFastaFile: null,
      selectedSequencesFastaFile: null,
      taxonomy: null,
      twoDimensionalData: new Array<number[]>(),
    };
  }

  public async componentDidMount() {
    const { numIterationsBeforeReRender, numSequencesToShow } = this.props;

    let taxonomy: null | { [key: string]: { [key: string]: string } } = null;
    Papa.parse('datasets/betalactamase_alignment/PSE1_NATURAL_TAXONOMY.csv', {
      complete(results) {
        console.log('csv results', results);

        taxonomy = results.data.reduce(
          (accumulator, seqTaxonomy) => {
            accumulator[seqTaxonomy.seq_name] = {
              class: seqTaxonomy.class,
              phylum: seqTaxonomy.phylum,
            };

            return accumulator;
          },
          {} as { [key: string]: { [key: string]: string } },
        );
      },
      download: true,
      header: true,
    });

    const fastaSequences = await fetchFastaFile('datasets/betalactamase_alignment/PSE1_natural_top5K_subsample.a2m');
    console.log('all fasta sequences:', fastaSequences);

    const randomFastaSequences = fastaSequences.getRandomSetOfSequences(numSequencesToShow);
    console.log('random fasta sequences:', randomFastaSequences);

    const integerAASequences = randomFastaSequences.getIntegerAASequences(); // faster than binary
    const t0 = performance.now();

    // WORKING FOR UMAP sequences dataset
    setTimeout(() => {
      const umap = new UMAP({ nComponents: 2, distanceFn: hammingDistance, minDist: 0.1, spread: 1 });

      const nEpochs = umap.initializeFit(integerAASequences);
      console.log(`UMAP wants to do ${nEpochs} epochs`);

      const stepUmapFn = (counter: number) => {
        if (counter % numIterationsBeforeReRender === 0 && counter <= nEpochs) {
          if (counter % 50 === 0) {
            console.log(counter + '::' + (performance.now() - t0) / 1000 + ' sec');
          }
          const embedding = umap.getEmbedding();
          if (counter === 0) {
            console.log('embedding:', embedding);
          }

          this.setState({
            allSequencesFastaFile: fastaSequences,
            selectedSequencesFastaFile: randomFastaSequences,
            taxonomy,
            twoDimensionalData: embedding,
          });
        }
        umap.step();
        if (counter <= nEpochs) {
          setTimeout(() => {
            stepUmapFn(counter + 1);
          });
        }
      };

      stepUmapFn(0);
    });

    /* // WORKING FOR UMAP scRNAseq dataset
    const { numIterationsBeforeReRender } = this.props;

    const data = await fetchMatrixData('datasets/hpc/full/tsne_matrix.csv');

    const { numIterationsBeforeReRender } = this.props;
    console.log(`UMAP got data (${data.length}, ${data[0].length})`);

    const umap = new UMAP({ nComponents: 2 });
    const nEpochs = umap.initializeFit(data);
    console.log(`UMAP wants to do ${nEpochs} epochs`);

    const stepUmapFn = (counter: number) => {
      if (counter % numIterationsBeforeReRender === 0 && counter <= nEpochs) {
        console.log(`UMAP on iteration ${counter}`);
        const embedding = umap.getEmbedding();
        this.setState({
          twoDimensionalData: embedding,
        });
      }
      umap.step();
      if (counter <= nEpochs) {
        setTimeout(() => {
          stepUmapFn(counter + 1);
        });
      }
    };

    setTimeout(() => {
      stepUmapFn(0);
    });*/
  }
  public render() {
    let plotType: string = 'scattergl';
    const dataX = new Float32Array(this.state.twoDimensionalData.map(datum => datum[0]));
    const dataY = new Float32Array(this.state.twoDimensionalData.map(datum => datum[1]));
    let dataZ: Float32Array | undefined;
    if (
      this.state.twoDimensionalData &&
      this.state.twoDimensionalData.length > 0 &&
      this.state.twoDimensionalData[0].length > 2
    ) {
      dataZ = new Float32Array(this.state.twoDimensionalData.map(datum => datum[3]));
      plotType = 'scatter3d';
    }

    let seqTitles = new Array<string>();
    if (this.state.selectedSequencesFastaFile) {
      seqTitles = this.state.selectedSequencesFastaFile.getSequenceObjects().map(seqObj => {
        return seqObj.title;
      });
    }

    let colors = new Array<string>();
    if (this.state.selectedSequencesFastaFile && this.state.taxonomy) {
      colors = this.state.selectedSequencesFastaFile.getSequenceObjects().map(seqObj => {
        if (this.state.taxonomy) {
          const taxProps = this.state.taxonomy[seqObj.title];
          const color = taxProps && taxProps.phylum ? this.getClassColor(taxProps.class) : null;
          if (color) {
            return color;
          }
        }

        return 'gray';
      });
    }

    // smart range setting
    if (this.xRange[0] > Math.min(...dataX)) {
      this.xRange[0] = Math.min(...dataX) - 5;
    }
    if (this.xRange[1] < Math.max(...dataX)) {
      this.xRange[1] = Math.max(...dataX) + 5;
    }
    if (this.yRange[0] > Math.min(...dataY)) {
      this.yRange[0] = Math.min(...dataY) - 5;
    }
    if (this.yRange[1] < Math.max(...dataY)) {
      this.yRange[1] = Math.max(...dataY) + 5;
    }

    return (
      <div>
        asdf
        <ComponentCard componentName={'UMAPVisualization'}>
          <PlotlyChart
            layout={{
              ...defaultPlotlyLayout,
              dragmode: 'zoom', // 'select',
              margin: {
                b: 20,
              },
              xaxis: { autorange: false, range: this.xRange },
              yaxis: { autorange: false, range: this.yRange },
            }}
            data={[
              {
                marker: {
                  color: colors,
                },
                mode: 'markers',
                text: seqTitles,
                type: 'scattergl', // scatter3d
                x: dataX,
                y: dataY,
                z: dataZ,
              },
            ]}
            showLoader={false}
          />
        </ComponentCard>
      </div>
    );
  }

  private getPhylumColor(phylumToFind: string | undefined): string | undefined {
    const phylumColors: { [key: string]: string } = {
      Actinobacteria: '#ff7f0e',
      Bacteroidetes: '#2ca02c',
      Firmicutes: '#8c564b',
      Proteobacteria: '#e377c2',
    };
    if (phylumToFind) {
      return phylumColors[phylumToFind];
    }

    return undefined;
  }

  private getClassColor(classToFind: string | undefined): string | undefined {
    const classColors: { [key: string]: string } = {
      Actinobacteria: '#ff7f0e',
      Alphaproteobacteria: '#1f77b4',
      Bacilli: '#2ca02c',
      Betaproteobacteria: '#d62728',
      Clostridia: '#9467bd',
      Gammaproteobacteria: '#17becf',
    };
    if (classToFind) {
      return classColors[classToFind];
    }
  }
}
