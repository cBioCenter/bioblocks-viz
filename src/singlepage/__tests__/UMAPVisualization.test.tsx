import { shallow } from 'enzyme';
import * as React from 'react';

import { Seq, SeqRecord } from '~bioblocks-viz~/data';
import { UMAPSequenceContainer, UMAPTranscriptionalContainer, UMAPVisualization } from '~bioblocks-viz~/singlepage';

describe('UMAPVisualization', () => {
  let sequence: Seq;
  let sequences: SeqRecord[];
  let taxonomyText: string;

  beforeEach(() => {
    jest.clearAllTimers();
    jest.useFakeTimers();
    sequence = new Seq('gatcctag');
    sequences = [new SeqRecord(sequence)];
    // Taxonomy text based off betalactamase_alignment - PSE1_NATURAL_TAXONOMY.csv
    taxonomyText = `seq_name,sequence,tax_id,superkingdom,phylum,genus,class,subphylum,family,order,species\n\
example_sequence,EVA-NERV,59846,Bacteria,Firmicutes,Paenibacillus,Bacilli,Paenibacillaceae,Bacillales,Paenibacillus chibensis`;
  });

  describe('UMAPSequenceContainer', () => {
    it('Should render when given an empty sequence.', () => {
      const wrapper = shallow(<UMAPSequenceContainer allSequences={[]} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('Should render when given a sequence.', () => {
      const wrapper = shallow(<UMAPSequenceContainer allSequences={sequences} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('Should handle a taxonomy update.', () => {
      const wrapper = shallow(<UMAPSequenceContainer allSequences={sequences} />);
      const expected = {
        example_sequence: {
          class: 'Bacilli',
          family: 'Bacillales',
          genus: 'Paenibacillus',
          order: 'Paenibacillus chibensis',
          phylum: 'Firmicutes',
          seq_name: 'example_sequence',
          sequence: 'EVA-NERV',
          subphylum: 'Paenibacillaceae',
          superkingdom: 'Bacteria',
          tax_id: '59846',
        },
      };
      wrapper.setProps({
        taxonomyText,
      });
      expect(wrapper.state('seqNameToTaxonomyMetadata')).toEqual(expected);
    });
  });

  describe('UMAPTranscriptionalContainer', () => {
    it('Should render when given an empty data matrix.', () => {
      const wrapper = shallow(<UMAPTranscriptionalContainer dataMatrix={[[]]} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('UMAPVisualization', () => {
    it('Should render when given an empty data matrix.', () => {
      const wrapper = shallow(<UMAPVisualization dataMatrix={[[]]} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('Should render when given some sample data.', () => {
      const dataMatrix: number[][] = new Array(30);
      for (let i = 0; i < 30; ++i) {
        dataMatrix[i] = new Array(30);
        for (let j = 0; j < 30; ++j) {
          dataMatrix[i][j] = i * 30 + j;
        }
      }
      const wrapper = shallow(<UMAPVisualization dataMatrix={dataMatrix} />);

      jest.runAllTimers();
      expect(wrapper.state('umapEmbedding')).toHaveLength(30);
    });
  });
});
