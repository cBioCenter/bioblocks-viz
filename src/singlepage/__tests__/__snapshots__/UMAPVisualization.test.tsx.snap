// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UMAPVisualization UMAPSequenceContainer Should render when given a sequence. 1`] = `
<UMAPVisualization
  allSequences={
    Array [
      SeqRecord {
        "alphabet": undefined,
        "annotations": Object {
          "dbxrefs": Array [],
          "description": "<unknown description>",
          "id": "<unknown id>",
          "letterAnnotations": Object {},
          "metadata": Object {},
          "name": "<unknown name>",
        },
        "back_transcribe": [Function],
        "binaryRepresentation": [Function],
        "complement": [Function],
        "determineAlphabet": [Function],
        "equal": [Function],
        "ignoredCharacters": Array [
          "-",
        ],
        "integerRepresentation": [Function],
        "isValidDNA": [Function],
        "isValidProtein": [Function],
        "isValidRNA": [Function],
        "lower": [Function],
        "reverse_complement": [Function],
        "sequence": "gatcctag",
        "subSequence": [Function],
        "toString": [Function],
        "transcribe": [Function],
        "translate": [Function],
        "upper": [Function],
      },
    ]
  }
  dataLabels={
    Array [
      undefined,
    ]
  }
  dataMatrix={
    Array [
      Array [
        0,
        1,
        2,
        3,
        3,
        2,
        1,
        0,
      ],
    ]
  }
  distanceFn={[Function]}
  errorMessages={Array []}
  labelCategory="class"
  numIterationsBeforeReRender={1}
  numSequencesToShow={2000}
  tooltipNames={
    Array [
      "<unknown name>",
    ]
  }
  xRange={
    Array [
      -20,
      20,
    ]
  }
  yRange={
    Array [
      -20,
      20,
    ]
  }
/>
`;

exports[`UMAPVisualization UMAPSequenceContainer Should render when given an empty sequence. 1`] = `
<UMAPVisualization
  allSequences={Array []}
  dataLabels={Array []}
  dataMatrix={Array []}
  distanceFn={[Function]}
  errorMessages={Array []}
  labelCategory="class"
  numIterationsBeforeReRender={1}
  numSequencesToShow={2000}
  tooltipNames={Array []}
  xRange={
    Array [
      -20,
      20,
    ]
  }
  yRange={
    Array [
      -20,
      20,
    ]
  }
/>
`;

exports[`UMAPVisualization UMAPTranscriptionalContainer Should render when given an empty data matrix. 1`] = `
<UMAPVisualization
  dataLabels={Array []}
  dataMatrix={
    Array [
      Array [],
    ]
  }
  distanceFn={[Function]}
  errorMessages={Array []}
  numIterationsBeforeReRender={1}
  numSamplesToShow={4000}
  tooltipNames={Array []}
  xRange={
    Array [
      -20,
      20,
    ]
  }
  yRange={
    Array [
      -20,
      20,
    ]
  }
/>
`;

exports[`UMAPVisualization UMAPVisualization Should render when given an empty data matrix. 1`] = `
<div>
  <ComponentCard
    componentName="UMAP"
    dockItems={
      Array [
        Object {
          "isLink": false,
          "text": "",
        },
      ]
    }
    frameHeight={0}
    frameWidth={0}
    headerHeight={20}
    height="525px"
    iconSrc="https://bioblocks.org/media/5d3f528028d9720a3428e4b2"
    isDataReady={false}
    isFramedComponent={false}
    isFullPage={false}
    menuItems={Array []}
    padding={0}
    showSettings={false}
    width="525px"
  >
    <PlotlyChart
      config={Object {}}
      data={
        Array [
          Object {
            "marker": Object {
              "color": Array [],
            },
            "mode": "markers",
            "text": undefined,
            "type": "scattergl",
            "x": Float32Array [
              NaN,
            ],
            "y": Float32Array [
              NaN,
            ],
          },
        ]
      }
      height="100%"
      layout={
        Object {
          "autosize": true,
          "dragmode": "zoom",
          "hovermode": "closest",
          "legend": Object {},
          "margin": Object {
            "b": 20,
          },
          "showlegend": false,
          "title": "",
          "xaxis": Object {
            "autorange": false,
            "range": Array [
              -20,
              20,
            ],
          },
          "yaxis": Object {
            "autorange": false,
            "range": Array [
              -20,
              20,
            ],
          },
        }
      }
      showLoader={true}
      width="100%"
    />
  </ComponentCard>
</div>
`;
